Fix international keyboard input for Linux input layer

fonic opened the pull request:
Fix keyboard input for international keyboard layouts/keymaps when using plain Linux input layer.

## Description
On platforms using the plain Linux input layer (e.g. OpenELEC/LibreELEC on Raspberry Pi), it is currently not possible to input characters other than ASCII on international keyboards.

This shortcoming has been present for a long time, probably from the very beginning, affecting lots of users. Unfortunately, the issue was never addressed correctly most likely due to a general misunderstanding of the concepts involved - from the bug reports below it is obvious that layouts, keymaps, input layers, virtual keyboard and physical keyboard are mixed up to the point where it is unclear which is what.

From reviewing debug logs on both a LibreELEC/Raspberry and a desktop Linux installation of Kodi, I discovered that unicode input was already working as far as the kernel providing the correct translated unicode character for a given keypress, but somehow this information got lost along the way. After a whole day of digging through the code, I managed to pinpoint the origin of the issue and found the fix to be **very** simple.

While at it, I also fixed two other issues (just for the sake of completeness):

- map mode key ('Alt Gr') to right Alt key like most OS do
- fix Euro currency sign

Related bug reports (probably many more):
https://github.com/OpenELEC/OpenELEC.tv/issues/108
https://github.com/OpenELEC/OpenELEC.tv/issues/739
https://github.com/OpenELEC/OpenELEC.tv/issues/1126
https://github.com/OpenELEC/OpenELEC.tv/issues/1661
https://github.com/OpenELEC/OpenELEC.tv/issues/1965
https://github.com/OpenELEC/OpenELEC.tv/issues/1986
https://github.com/OpenELEC/OpenELEC.tv/issues/2017
https://github.com/OpenELEC/OpenELEC.tv/issues/2094
https://github.com/OpenELEC/OpenELEC.tv/issues/3943
https://github.com/OpenELEC/OpenELEC.tv/issues/4282
http://openelec.tv/forum/124-raspberry-pi/75575-umlaute-not-working-logitec-k400-openelec-5-0-3
http://openelec.tv/forum/105-keyboards/24435-german-keyboard-layout
https://forum.libreelec.tv/thread/5757-raspberry-pi2-le-special-characters-not-work-from-keyboard/
https://forum.libreelec.tv/thread/3384-problems-with-ü-ö-ä-and-ß/
https://forum.libreelec.tv/thread/2107-norwegian-keyboard-layout-norwegian-letters/
https://forum.libreelec.tv/thread/1521-special-characters-in-spanish-keyboard-raspberry-pi-3/

## Motivation and Context
Enable users with keyboard layouts other than _en-US_ to input all possible characters (or, more precisely, all characters supported by Kodi) on platforms using the plain Linux input layer.

## How Has This Been Tested?
Tested successfully with **LibreELEC** (compiled from Git snapshot of 'master' branch as of 09/02/17) on **Raspberry Pi 3**, German keyboard attached via USB and keymap set to 'de-latin1-nodeadkeys'. With the patch applied, I was able to input the characters **üöäÜÖÄß€**, without the patch pressing those keys does nothing.

Please note that the fix is universal, i.e. it should work with all keymaps/layouts.

As far as I can tell, the fix has no side-effects (needs to be verified by others!).

## Screenshots (if appropriate):

## Types of change
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

stefansaraev commented on xbmc/input/linux/LinuxInputDevices.cpp:
```diff
@@ -370,7 +370,7 @@ int CLinuxInputDevice::KeyboardGetSymbol(unsigned short value)
       return XBMCK_BACKSPACE;
     case 0xa4:
-      return XBMCK_EURO; /* euro currency sign */
+      return 0x20ac; /* euro currency sign */
```
should this be updated in xbmc/input/XBMC_keysym.h instead ?

fonic replied:
Certainly possible (I actually considered this), but since input of the Euro sign already worked flawlessly on other platforms (tested on desktop Linux), I decided not to mess with it to avoid possible side-effects.

graham8 replied:
+1 for coding it here. 0x20ac is a unicode value, not a XBMC Key.

vkravets commented:
@fonic Today finally compile Libreelec with your path and unfortunately it's working partially... I'm able to input non latin symbols by shifting Alt but... it's shown as not international symbols.

For my case I've used Libreelec on S905 box with loadkeys flow.
![screenshot000](https://user-images.githubusercontent.com/985336/30135203-a98e346e-9362-11e7-8acc-becc19ad779e.png)

In the settings was select ru_win layout
which can be found here https://fossies.org/linux/kbd/data/keymaps/i386/qwerty/ru_win.map

vkravets commented:
```
00:17:27.596 T:4099166208   DEBUG: Keyboard: scancode: 0x64, sym: 0x0139, unicode: 0x0000, modifier: 0x200
00:17:27.596 T:4099166208   DEBUG: OnKey: alt-165 (0x4f0a5, obc-323494) pressed, trying keyboard action f0a5
00:17:27.596 T:4099166208   DEBUG: OnKey: alt-165 (0x4f0a5, obc-323494) pressed, action is
00:17:27.822 T:4099166208   DEBUG: Keyboard: scancode: 0x20, sym: 0x00e2, unicode: 0x00e2, modifier: 0x200
00:17:27.822 T:4099166208   DEBUG: OnKey: alt-165 (0x4f0a5, obc-323494) pressed, trying keyboard action f100
00:17:27.956 T:4099166208   DEBUG: Keyboard: scancode: 0x20, sym: 0x00e2, unicode: 0x00e2, modifier: 0x200
00:17:28.322 T:4099166208   DEBUG: Previous line repeats 1 times.
00:17:28.322 T:4099166208   DEBUG: OnKey: alt-165 (0x4f0a5, obc-323494) pressed, trying keyboard action f100
00:17:28.455 T:4099166208   DEBUG: Keyboard: scancode: 0x20, sym: 0x00e2, unicode: 0x00e2, modifier: 0x200
00:17:28.656 T:4099166208   DEBUG: Previous line repeats 1 times.
00:17:28.656 T:4099166208   DEBUG: OnKey: alt-165 (0x4f0a5, obc-323494) pressed, trying keyboard action f100
00:17:28.790 T:4099166208   DEBUG: Keyboard: scancode: 0x20, sym: 0x00e2, unicode: 0x00e2, modifier: 0x200
00:17:29.290 T:4099166208   DEBUG: Keyboard: scancode: 0x2d, sym: 0x00f7, unicode: 0x00f7, modifier: 0x200
00:17:29.290 T:4099166208   DEBUG: OnKey: alt-165 (0x4f0a5, obc-323494) pressed, trying keyboard action f100
00:17:29.290 T:4099166208   DEBUG: Keyboard: scancode: 0x2e, sym: 0x00f1, unicode: 0x00f1, modifier: 0x200
00:17:29.290 T:4099166208   DEBUG: OnKey: alt-165 (0x4f0a5, obc-323494) pressed, trying keyboard action f100
00:17:29.323 T:4099166208   DEBUG: Keyboard: scancode: 0x2d, sym: 0x00f7, unicode: 0x00f7, modifier: 0x200
00:17:29.390 T:4099166208   DEBUG: Keyboard: scancode: 0x2e, sym: 0x00f1, unicode: 0x00f1, modifier: 0x200
00:17:29.722 T:4099166208   DEBUG: Previous line repeats 1 times.
00:17:29.722 T:4099166208   DEBUG: OnKey: alt-165 (0x4f0a5, obc-323494) pressed, trying keyboard action f100
00:17:29.890 T:4099166208   DEBUG: Keyboard: scancode: 0x2e, sym: 0x00f1, unicode: 0x00f1, modifier: 0x200
00:17:30.123 T:4099166208   DEBUG: Previous line repeats 1 times.
00:17:30.123 T:4099166208   DEBUG: OnKey: alt-165 (0x4f0a5, obc-323494) pressed, trying keyboard action f100
00:17:30.254 T:4099166208   DEBUG: Keyboard: scancode: 0x2e, sym: 0x00f1, unicode: 0x00f1, modifier: 0x200
00:17:31.887 T:4099166208   DEBUG: Keyboard: scancode: 0x1e, sym: 0x00f4, unicode: 0x00f4, modifier: 0x200
00:17:31.887 T:4099166208   DEBUG: OnKey: alt-165 (0x4f0a5, obc-323494) pressed, trying keyboard action f100
00:17:32.022 T:4099166208   DEBUG: Keyboard: scancode: 0x1e, sym: 0x00f4, unicode: 0x00f4, modifier: 0x200
00:17:32.287 T:4099166208   DEBUG: Keyboard: scancode: 0x1f, sym: 0x00fb, unicode: 0x00fb, modifier: 0x200
00:17:32.287 T:4099166208   DEBUG: OnKey: alt-165 (0x4f0a5, obc-323494) pressed, trying keyboard action f100
00:17:32.433 T:4099166208   DEBUG: Keyboard: scancode: 0x1f, sym: 0x00fb, unicode: 0x00fb, modifier: 0x200
00:17:32.620 T:4099166208   DEBUG: Previous line repeats 1 times.
00:17:32.621 T:4099166208   DEBUG: OnKey: alt-165 (0x4f0a5, obc-323494) pressed, trying keyboard action f100
00:17:32.758 T:4099166208   DEBUG: Keyboard: scancode: 0x1f, sym: 0x00fb, unicode: 0x00fb, modifier: 0x200
00:17:32.857 T:4099166208   DEBUG: Keyboard: scancode: 0x64, sym: 0x0139, unicode: 0x0000, modifier: 0x0
00:17:34.322 T:4099166208   DEBUG: Keyboard: scancode: 0x1d, sym: 0x0132, unicode: 0x0000, modifier: 0x40
00:17:34.322 T:4099166208   DEBUG: OnKey: ctrl-leftctrl (0x1f0d0) pressed, trying keyboard action f0d0
00:17:34.322 T:4099166208   DEBUG: OnKey: ctrl-leftctrl (0x1f0d0) pressed, action is
00:17:34.555 T:4099166208   DEBUG: Keyboard: scancode: 0x25, sym: 0x006b, unicode: 0x006b, modifier: 0x40
00:17:34.556 T:4099166208   DEBUG: OnKey: ctrl-k (0x1f04b) pressed, trying keyboard action f16b
00:17:34.689 T:4099166208   DEBUG: Keyboard: scancode: 0x25, sym: 0x006b, unicode: 0x006b, modifier: 0x40
00:17:35.189 T:4099166208   DEBUG: Previous line repeats 1 times.
00:17:35.189 T:4099166208   DEBUG: OnKey: ctrl-k (0x1f04b) pressed, trying keyboard action f16b
00:17:35.322 T:4099166208   DEBUG: Keyboard: scancode: 0x25, sym: 0x006b, unicode: 0x006b, modifier: 0x40
00:17:36.122 T:4099166208   DEBUG: Keyboard: scancode: 0x26, sym: 0x006c, unicode: 0x006c, modifier: 0x40
00:17:36.122 T:4099166208   DEBUG: OnKey: ctrl-l (0x1f04c) pressed, trying keyboard action f16c
00:17:36.289 T:4099166208   DEBUG: Keyboard: scancode: 0x26, sym: 0x006c, unicode: 0x006c, modifier: 0x40
00:17:36.756 T:4099166208   DEBUG: Keyboard: scancode: 0x1d, sym: 0x0132, unicode: 0x0000, modifier: 0x0
00:17:37.455 T:4099166208   DEBUG: Keyboard: scancode: 0x38, sym: 0x0134, unicode: 0x0000, modifier: 0x100
00:17:37.456 T:4099166208   DEBUG: OnKey: alt-leftalt (0x4f0d4) pressed, trying keyboard action f0d4
00:17:37.456 T:4099166208   DEBUG: OnKey: alt-leftalt (0x4f0d4) pressed, action is
00:17:37.822 T:4099166208   DEBUG: Keyboard: scancode: 0x25, sym: 0x00eb, unicode: 0x00eb, modifier: 0x100
00:17:37.822 T:4099166208   DEBUG: OnKey: alt-164 (0x4f0a4, obc-323493) pressed, trying keyboard action f100
00:17:37.922 T:4099166208   DEBUG: Keyboard: scancode: 0x25, sym: 0x00eb, unicode: 0x00eb, modifier: 0x100
00:17:38.789 T:4099166208   DEBUG: Keyboard: scancode: 0x33, sym: 0x00e1, unicode: 0x00e1, modifier: 0x100
00:17:38.789 T:4099166208   DEBUG: OnKey: alt-164 (0x4f0a4, obc-323493) pressed, trying keyboard action f100
00:17:38.957 T:4099166208   DEBUG: Keyboard: scancode: 0x33, sym: 0x00e1, unicode: 0x00e1, modifier: 0x100
00:17:39.855 T:4099166208   DEBUG: Keyboard: scancode: 0x34, sym: 0x00fe, unicode: 0x00fe, modifier: 0x100
00:17:39.856 T:4099166208   DEBUG: OnKey: alt-164 (0x4f0a4, obc-323493) pressed, trying keyboard action f100
00:17:39.990 T:4099166208   DEBUG: Keyboard: scancode: 0x34, sym: 0x00fe, unicode: 0x00fe, modifier: 0x100
00:17:40.153 T:4099166208   DEBUG: Keyboard: scancode: 0x38, sym: 0x0134, unicode: 0x0000, modifier: 0x0
00:17:41.622 T:4099166208   DEBUG: Keyboard: scancode: 0x6c, sym: 0x0112, unicode: 0x0000, modifier: 0x0
00:17:41.622 T:4099166208   DEBUG: OnKey: down (0xf081) pressed, trying keyboard action 4
00:17:41.755 T:4099166208   DEBUG: Keyboard: scancode: 0x6c, sym: 0x0112, unicode: 0x0000, modifier: 0x0
```

Any thoughts how to fix this?

fonic commented:
@vkravets:
I don't believe the `ru_win` keymap will work as it does not contain any Cyrillic characters.

Try the following:

- On your desktop Linux machine:
  `zcat /usr/share/keymaps/i386/qwerty/ru.map.gz > /tmp/ru.map`
  `loadkeys -b /tmp/ru.map > /tmp/ru.bmap`
  `scp /tmp/ru.bmap <ip_of_libreelec_box>/tmp/ru.bmap`
- On your LibreELEC box via SSH:
  `loadkmap < /tmp/ru.bmap`

I can't test this because `loadkeys -b /tmp/ru.map > /tmp/ru.bmap` fails on my machine complaining about unknown Cyrillic symbols, but it might work for you.

graham8 commented on xbmc/input/linux/LinuxInputDevices.cpp:
```diff
@@ -444,6 +444,7 @@ XBMCMod CLinuxInputDevice::UpdateModifiers(XBMC_Event& devt)
+    case XBMCK_MODE: modifier = XBMCKMOD_RALT; break;
     default: break;
   }
 
```
It would be more logical to map KEY_RIGHTALT to XBMCK_RALT.

fonic replied:
No - there is not right Alt key on keyboards featuring the mode key.

fonic replied:
Forget that, I got that wrong. You're right, it should be possible to change:
`{ KEY_RIGHTALT , XBMCK_MODE },`
to
`{ KEY_RIGHTALT , XBMCK_RALT },`
in _xbmc/input/linux/LinuxInputDevices.cpp_.

graham8 commented:
> Please note that the fix is universal, i.e. it should work with all keymaps/layouts.
This a major step forward, thanks, but it will not work with keymaps returning windows codepage values or ISO8859 values other than 'latin 1', while the kodi charset is 'default' (unicode, presumably).  I've not managed to get non-unicode keymaps to play nicely with any setting for the charmaps.  Users need to stick to unicode keymaps.

There is also no support for keymaps which toggle between two different charsets (Greek, Russian, ...).  I'm working on that.

samnazarko commented:
>There is also no support for keymaps which toggle between two different charsets (Greek, Russian, ...). I'm working on that.

I'm not sure whether there will be a way of doing this that can be upstreamed; particularly as there seems to be differences in how keymaps are handled depending on the underlying display mechanism (X11 / GLES etc). 

@graham8 For OSMC, we would probably want to keep the system keyboard layout and Kodi's keyboard  layout in sync. For this, we add a setting callback for this in CSettings. In CApplication::Create we should set Kodi's keyboard layout to that of the system and handle setting changes in Kodi properly so we sync them back to the system.

Unfortunately, I think this may be tricky to map Kodi keyboard layouts to system keyboard layouts. Looking at console-data, the config (ran on dpkg-reconfigure ...) seems quite large:

```
dpkg -I console-data_1.12-5_all.deb 
 new debian package, version 2.0.
 size 1149680 bytes: control archive=188319 bytes.
   29964 bytes,  1285 lines   *  config               #!/usr/bin/perl
     892 bytes,    21 lines      control              
   48610 bytes,   632 lines      md5sums              
     726 bytes,    27 lines   *  postinst             #!/bin/sh
     206 bytes,     8 lines   *  postrm               #!/bin/sh
  431271 bytes,  2848 lines      templates            
```

That's a lot of setup for just configuring the keyboard layout.

graham8 commented:
> I'm not sure whether there will be a way of doing this that can be upstreamed; 

You will see shortly there is :)

vkravets commented:
@graham8 Wow, thanks a lot, working like a charm )
@fonic also ru.map working well

Thanks guys

vkravets commented:
@graham8 Could you please check, but it seems in non latin mode space is not working for keyboard mode.

# Open any keyboard input with onscreen keyboard
# Switch to non latin layout 
# Try to press space

In my case using ru.bmap which was mentioned above space is not working in non latin layout under kodi keyboard mode

vkravets commented:
Any updates? It will be great if it will be merged to master...

graham8 commented:
Well my PR has been merged in OSMC ready for the next release.  I gave up on a PR here when I couldn't get my head round the git structure. Meanwhile, this has cropped up: https://github.com/xbmc/xbmc/pull/13176

If it gets traction, the thing will have to be re-worked.  I can't tell from the comments how likely it is to be adopted.  And TBH don't understand why it's necessary.

garbear commented:
lrusak and I decided to make a binary add-on for libinput, and this should be usable on all platforms. If this is the case then the code here can likely be removed.

graham8 commented:
Is there somewhere in the docs where the concept of a 'binary add-on' is explained? 'Add-on' implies something optional to the basic application. 'Library' implies something available to applications other than kodi.  To an outsider like me, neither if these seem appropriate to the case.

notspiff commented:
it's shared library loaded by kodi. joysticks are already done in a binary add-on, https://github.com/xbmc/peripheral.joystick
since the interfaces are already there, it's very appropriate.

garbear commented:
You make a good point though, libinput offers many advantages but if the builder doesn't compile binary add-ons the user will lose input on all linux systems. Maybe we opportunistically use the libinput add-on, and fall back to this code if it's not present?

graham8 commented:
So two lots of code to maintain? I don't think I'd go for that if I was in charge.

garbear commented:
Then let's drop OS input and make the libinput add-on mandatory (kinda like [this](https://github.com/xbmc/xbmc/commit/e834745) but without the "optional")

fonic committed:
Fix international keyboard input for Linux input layer

Fix keyboard input for international keyboard layouts/keymaps when using plain Linux input layer (e.g. LibreELEC on Raspberry Pi):
- allow input of unicode characters
- map mode key ('Alt Gr') to right Alt key
- fix euro currency sign

wsnipex commented:
I'm strongly against moving core functionality such as input to an addon

graham8 commented:
Indeed.  Mandatory add-on sounds like an oxymoron.

notspiff commented:
Skin, language wonder why you are calling them bad things :(

It is pure organizational stuff. We have an interface for this. We would build it with core. It is the same as any other dll, just using the interface.

graham8 commented:
OK, if 'add-on' is just poor semantics, I bow to your better knowledge. Will wait and see how it turns out.

lrusak commented:
@wsnipex the libinput or will require adding dependencies for libinput, xkbcommon, mtdev, and possibly evdev. if I don't have to add those depends into core that would be nice.

If we can extend the CPeripherals to support keyboards, mice, and touchscreens that would be the best route then using the antiquated winevents stuff. Also bear in mind that this only effects RPi, AML, and GBM platforms.

wsnipex commented:
@spiff as long as it's build with core, it's ok for me. What must never happen is that you build kodi core and are left without input handling.
@lrusak you still need to add those deps somewhere though.

notspiff commented:
yes, we certainly can't ship an application without input support :) but by doing it like this we gain the ability to update support for a library that is far from stable between releases, and we get modularity through extensibility.

Rechi commented:
Is this obsolete with #13176?

graham8 commented:
Yes

samnazarko commented:
+1 from osmc

Sent from a mobile device, so please excuse the brevity of this reply

On 3 May 2018, at 22:13, Graham <notifications@github.com<mailto:notifications@github.com>> wrote:


Yes

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/xbmc/xbmc/pull/12762#issuecomment-386438315>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AA47LOxOaVooM6TYpOp90gm231qQ6D_qks5tu3LbgaJpZM4PK1ox>.
